### DISCUSSION TITLE
Working with remote files in EIN

### DISCUSSION THREAD

**Discussion by gzagatti** [emacs-lsp/lsp-mode#3355] [Q&A] [answered]
I am working with remote files served by a Jupyter server using [EIN](https://github.com/millejoh/emacs-ipython-notebook). However, I am having trouble opening multiple python files from the same server.

My issue goes as following:

1. Assume my current working directory is `/my/path`
1. Connect to a jupyter server `https://myserver.com` using `ein:notebooklist-open`
2. Select a file to open. It looks like EIN modifies the buffer name according to a set of rules described [here](https://github.com/millejoh/emacs-ipython-notebook/blob/master/lisp/ein-file.el). The basic idea is that if the file in the server is located in `remote-dir/path`, the buffer is renamed to `/my/path/'ein:https://myserver.com:remote-dir/path`.
3. LSP will then ask about the root of the project. Since `'ein:https://myserver.com:remote-dir/` should not be mirrored in the local file system I manually select the project root with `I` and set it to `/my/path`. LSP then loads without error and LSP does not interfere with the write hook that is implemented by EIN on the link pointed in (2).
4. Now, I want to open a second file `remote-dir/another-path`. Now, LSP tries to save the file to `/my/path/'ein:https://myserver.com:remote-dir/another-path`. Every time I try to save the file, it keeps complaining that the path  `/my/path/'ein:https://myserver.com:remote-dir/another-path` does not exist, which is true and intentional. If I say I do not want to write the file, I get an error message `error in process sentinel: Canceled`.

My current working solution which is far from ideal consists of the following:

1. When I try to open a second file, I refuse to save the file in `/my/path/'ein:https://myserver.com:remote-dir/another-path`
2. Then, I manually add the hook  `(add-hook 'write-contents-functions 'ein:content-file-save)`
3. The LSP server gets activated on the second file using the LSP server of the first file since they now share the same root. 
3. Save now works as expected, sending the contents to the jupyter server.

How can I modify the behavior described above such that the following happens:

1. When opening a buffer that has the `ein` name pattern, LSP sets the working directory to the current working directory `/my/path`.
2. It lets EIN complete its job and add the required hook.

Any help would be appreciated. Based on this [FAQ](https://emacs-lsp.github.io/lsp-mode/page/faq/#how-do-i-force-lsp-mode-to-forget-the-workspace-folders-for-multi-root), my intuition is that I should modify `lsp` via `advice-add` but I am not sure how to do that.

**Comment by yyoncho** [comment, marked as answer]
Maybe the first thing to change is `lsp-auto-touch-files`? We automatically try to save the file on the disk(some language servers tend to fail if the file is not on the disk).

**Reply by gzagatti** [reply]
Thanks! That's what I needed. I'm glad the solution was so simple.
